src/utils/snap_utils.cpp

```diff
@@ -15,27 +15,32 @@
  *
  */
 
+#include <multipass/exceptions/snap_environment_exception.h>
 #include <multipass/snap_utils.h>
 
 #include <QFileInfo>
 
+namespace mp = multipass;
 namespace mu = multipass::utils;
 
-bool mu::is_snap()
-{
-    // Decide if Snap based on if $SNAP env var is set.
-    // TODO: can this be better?
-    return !snap_dir().isEmpty();
-}
-
 QByteArray mu::snap_dir()
 {
     auto snap_dir = qgetenv("SNAP");                         // Inside snap, this can be trusted.
+    if (snap_dir.isEmpty())
+    {
+        throw mp::SnapEnvironmentException("SNAP");
+    }
+
     return QFileInfo(snap_dir).canonicalFilePath().toUtf8(); // To resolve any symlinks
 }
 
 QByteArray mu::snap_common_dir()
 {
     auto snap_common = qgetenv("SNAP_COMMON");                  // Inside snap, this can be trusted
+    if (snap_common.isEmpty())
+    {
+        throw mp::SnapEnvironmentException("SNAP_COMMON");
```

Comment:
I suppose both `SNAP` and `SNAP_COMMON` should be guaranteed when running in a snap (right?), otherwise, it would be quite easy to disable apparmor profiles... But the opposite isn't true: we can have any of those defined when not in a snap.

By treating `SNAP` and `SNAP_COMMON` separately, we're relying on them being either both defined or both not defined. If we run in an environment where one of them is defined but not the other, we'll be doing some things as in a snap and some others as not in a snap. The situation was slightly different before this PR, because we would always gate on `SNAP` being defined (via `is_snap`), even when reading `SNAP_COMMON`.

Reply:
> But the opposite isn't true: we can have any of those defined when not in a snap.

Why do we care about this scenario?  This implies that someone is running from source (like us on the project) or has packaged Multipass in some other format **and** set `SNAP` and/or `SNAP_COMMON` when `multipassd` is running.  IMO, that is not a supported scenario and the user is on their own and we shouldn't add complexity to guard against that.

Reply:
I am not sure we need to do anything, but I wanted to point it out and get feedback. 

I was thinking that users could have vars with the same name in their environment, even on other platforms. That is unlikely in the daemon, but not impossible. And utils can be used in clients too, and in cross-platform code, even if that is not the case today. Of course such variables would already cause problems without this PR, but now there is a slight difference which I wanted to make sure of.

I suppose we can say that we don't support those variables outside of snaps and that we require both to be either defined or undefined simultaneously. Ideally that would involve some mention in documentation.

Reply:
Your points here are duly noted :grin:  I think we should keep it in mind, but leave it as is for now.

Reply:
Hey @ricab,

After thinking about this some more, I took your feedback and added a new function in the `snap_utils` to verify that `SNAP_NAME` == `multipass` before trying to get either snap directory environment variables.  This should at least add another level of checks on making sure we are running as a snap or not.  Let me know what you think.

Reply:
Thanks, yes I think that makes sense, as it adds another way in which we catch bad snap-like environments early, so certainly an improvement.

One way to enforce all three vars would be a class that initialized everything in the ctor and threw when any of them was not defined. Something like:

```cpp

class SnapInfo
{
public:
  SnapInfo() 
    : snap_dir{checked_getenv(...)}
    , snap_common(checked_getenv(...)}
  {
    checked_getenv("SNAP_NAME");
  }
  
  const string& snap_dir() const;
  const string& snap_common() const;

private:
  string checked_getenv(const string& var_name)
  {
    if (string val = getenv(var_name); val)
      return val;
    throw ...;
  }

  string snap_dir;
  string snap_common;
}

// client would do
try
{
  auto snap_dir = mu::SnapInfo{}.snap_dir().
  // or
  auto snap_dir = mu::SnapInfo::instance().snap_dir().
}
catch...
```

We'd be reading 3 vars each time though, instead of the current 2. The extra reads could be saved by making it a singleton. It wouldn't require mocking since we can mock everything we need at the environment level (and the singleton can be reset). 

But existing tests would need updating, so it may be more work than it's worth. I leave it up to you.

Reply:
Hey @ricab,

Thanks for the feedback!  I may just go the singleton route here since I think the calls to `getenv()` every time is redundant and can be streamlined in a singleton class as you say.

Reply:
Ok, I've been working on a singleton approach and it is not a trivial change, particularly around the tests.  *Any* test that uses the snap stuff will have to account for resetting the singleton in its test fixture, which many of the tests don't have one.

Although I think the singleton approach is the cleanest, it adds quite a bit of complexity for little gain.  I think leaving it how it is in this PR is the best use of time.

Reply:
Yes, I suspected that. OK, fine by me.

Reply:
Hi @townsend2010, I started looking at the code again and realized that what you have now still requires repeated guards in each test.

```c++
    QTemporaryDir snap_dir;
    mpt::SetEnvScope env("SNAP_COMMON", snap_dir.path().toUtf8());
    mpt::SetEnvScope env2("SNAP_NAME", snap_name);
```

This code is repeated each time and IMO it should be replaced with a fixture setup anyway...

Reply:
We do different things with the env vars in each test, so I don't think a test fixture would help here.

Reply:
IIUC, outside snap-utils we need only 2 situations: snap and non-snap. IOW: 

```c++
EXPECT_CALL(mock_snap_detector, get_[common_]dir).WillOnce(Return(...)|Throw(...));
```

Only snap detection tests themselves need further granularity. A singleton would rid us of a lot of boiler plate in tests, separate them into better *units*, read the environment O(1) times, instead of O(n), and enforce consistency among all 3 vars.

That said, I agree it would be a significant additional step that we can leave for a future PR.